api/storage/deployment_storage.go

```diff
@@ -83,3 +89,91 @@ func (d *deploymentStorage) GetFirstSuccessModelVersionPerModel() (map[models.ID
 func (d *deploymentStorage) Delete(modelID models.ID, versionID models.ID) error {
 	return d.db.Where("version_id = ? AND version_model_id = ?", versionID, modelID).Delete(models.Deployment{}).Error
 }
+
+// OnDeploymentSuccess updates the new deployment status to successful on DB and update all previous successful deployment status for that version endpoint to terminated.
+func (d *deploymentStorage) OnDeploymentSuccess(newDeployment *models.Deployment) error {
+	if newDeployment.ID == 0 {
+		return fmt.Errorf("newDeployment.ID must not be empty")
+	}
+
+	if !newDeployment.IsSuccess() {
+		return fmt.Errorf("newDeployment.Status must be running or serving")
+	}
+
+	tx := d.db.Begin()
+	if tx.Error != nil {
+		return fmt.Errorf("failed to begin transaction: %w", tx.Error)
+	}
+
+	var err error
+	defer func() {
+		if err != nil {
+			tx.Rollback()
+		} else {
+			tx.Commit()
+		}
+	}()
+
+	var deployments []*models.Deployment
+	err = tx.Where("version_model_id = ? AND version_id = ? AND version_endpoint_id = ? AND status IN ('pending', 'running', 'serving')",
```

Comment:
this is updating all deployment under the version right? or was this comment before update 😅

Reply:
Ah perhaps my phrasing wasn't clear. I was specifically referring to "pending" deployments. Currently the code is updating pending/serving/running deployments. But "pending" is a special scenario for which I am not sure what the correct handling should be.

> The API is expected to prevent further redeployment / undeployment attempts if there are current pending deployments, right? This is currently blocked from the UI but we should also block this from the API (there is a card for this in the backlog) because otherwise, the K8s resources from the "pending" deployment are not cleaned up.

> Alternatively, we should handle the clean up of the K8s resources for the pending deployment in the undeploy/redeploy endpoint flow - in this case, it makes sense to handle the pending deployment here, as you've done.

> What is the desired behavior of endpoint update / delete when a deployment is already in progress?

Reply:
I would assume `pending` is only applicable to the latest/ongoing deployment, since its not a terminal state. 

If there was past hanging deployment that got stuck in pending, then it should be updated to `failed` since it never succeeded

Reply:
> If there was past hanging deployment that got stuck in pending, then it should be updated to failed since it never succeeded

Makes sense. So, by the time this code is reached, we shouldn't have any "pending" deployments anymore (unless something was stuck in the past and not cleaned up properly; but this shouldn't be masked by updating the pending status to terminated in this block of code). So, TL;DR: we should drop "pending" from the query here.

Reply:
Late to the party but yeah I agree `pending` should be dropped from this query once the we block further redeployment/undeployment of pending deployments on the API server. 

Just one more not so important question, is it overkill to be updating _all_ deployments that have the `running`/`serving` status to terminated? Or should only the existing _current_ deployment's status be updated (if it is indeed `running` or `serving`)? Granted, this doesn't change the end result and will still be super safe but was just wondering if this blanket update might potentially hide some unexpected behaviour (say for some reason we have deployments that are somehow `running` or `serving` but are not the current deployment - would updating their status in the DB would hide some potential underlying problem?)

Reply:
> is it overkill to be updating all deployments that have the running/serving status to terminated?

The problem is, we are not actually tracking the "current" deployment record anywhere - it's identified to be a deployment that's in:
* running / serving state
* (OR) latest failed deployment

Only the first one needs to be updated for this workflow, to terminated.

> if this blanket update might potentially hide some unexpected behaviour (say for some reason we have deployments that are somehow running or serving but are not the current deployment

This is in fact the case today. But it should go away after the DB migration captured in this PR.

Reply:
@ariefrahmansyah perhaps this comment was overlooked. Could you review / address this too ?

Reply:
I'll remove the 'pending' state on both `OnDeploymentSuccess` and `Undeploy` methods as I agree that 'pending' state should be handled differently.

> is it overkill to be updating all deployments that have the running/serving status to terminated?

Once DB migration executed, there should be at most one running or serving deployment status for a given model version endpoint. So the query will only update at most one row.

> if this blanket update might potentially hide some unexpected behaviour (say for some reason we have deployments that are somehow running or serving but are not the current deployment - would updating their status in the DB would hide some potential underlying problem?)

This should be fine. As suggested in your question, running or serving deployments that not the current revision are safe to be updated to terminated status as it's not being use everywhere -- only on deployment history log UI.